### PR TITLE
firedoor deconstruction

### PR DESCRIPTION
**Pull request by ghost**
:cl:
* rscadd: Firedoors can now be deconstructed, weld them then use a socket wrench on them
* rscadd: Adds firedoor frames you can use to construct firedoors

carp hut beta 2017-03-20 234136

### CODE REVIEW

**Comment by FudgePucker** [comment]
can you make the frames out of metal?
If so it might be griffable to just place like 20 of them in the hallways then flip the fire alarm. Socket wrenches are not everyday items you see on the station, and if you have access you can always just space all two of the only ones on the station.

**Comment by kernikov** [comment]
Wouldn't that be as easy as adding a socket wrench step to construction, as long as the door isn't dense yet.

**Comment by ghost** [comment]
Will do.

Same cost as a airlock frame?

**Comment by ghost** [comment]
Wait, shit, that was a question.

@FudgePucker You can't make more of them as of right now.

File: code/game/machinery/doors/firedoor.dm [hunk 1]
```diff
@@ -501,7 +512,70 @@ var/global/list/alert_overlays_global = list()
...
+	name = "firedoor frame"
+	icon = 'icons/obj/doors/DoorHazard.dmi'
+	icon_state = "firedoor_frame"
+	inhand_states = list("left_hand" = 'icons/mob/in-hand/left/electronics.dmi', "right_hand" = 'icons/mob/in-hand/right/electronics.dmi')
+	desc = "The frame for a firedoor, strangely easy to set up considering its application."
+
+/obj/item/firedoor_frame/attack_self(mob/user)
+	construct_firedoor(user)
+
+/obj/item/firedoor_frame/proc/construct_firedoor(mob/user)
```
Comment (PJB3005) [inline code/game/machinery/doors/firedoor.dm]: ...Why not put this in the `attack_self` directly?

File: code/game/machinery/doors/firedoor.dm [hunk 2]
```diff
@@ -501,7 +512,70 @@ var/global/list/alert_overlays_global = list()
...
+	if(!user || !src)
+		return 0
+	if(!istype(user.loc,/turf))
```
Comment (PJB3005) [inline code/game/machinery/doors/firedoor.dm]: `!isturf()`

File: code/game/machinery/doors/firedoor.dm [hunk 3]
```diff
@@ -501,7 +512,70 @@ var/global/list/alert_overlays_global = list()
...
+	icon = 'icons/obj/doors/DoorHazard.dmi'
+	icon_state = "firedoor_frame"
+	inhand_states = list("left_hand" = 'icons/mob/in-hand/left/electronics.dmi', "right_hand" = 'icons/mob/in-hand/right/electronics.dmi')
+	desc = "The frame for a firedoor, strangely easy to set up considering its application."
+
+/obj/item/firedoor_frame/attack_self(mob/user)
+	construct_firedoor(user)
+
+/obj/item/firedoor_frame/proc/construct_firedoor(mob/user)
+	if(!user || !src)
```
Comment (PJB3005) [inline code/game/machinery/doors/firedoor.dm]: `!src` is literally impossible.

File: code/game/machinery/doors/firedoor.dm [hunk 4]
```diff
@@ -501,7 +512,70 @@ var/global/list/alert_overlays_global = list()
...
+		return 0
+	if(!istype(user.loc,/turf))
+		return 0
+	if(!user.IsAdvancedToolUser())
+		to_chat(user, "<span class='warning'>You don't have the dexterity to do this!</span>")
+		return 0
+	switch(alert("firedoor construction", "Would you like to construct a full tile firedoor or one direction?", "One Direction", "Full Firedoor", "Cancel", null))
+		if("One Direction")
+			if(!src)
```
Comment (PJB3005) [inline code/game/machinery/doors/firedoor.dm]: Not possible.

File: code/game/machinery/doors/firedoor.dm [hunk 5]
```diff
@@ -501,7 +512,70 @@ var/global/list/alert_overlays_global = list()
...
+				return 1
+			if(src.loc != user)
```
Comment (PJB3005) [inline code/game/machinery/doors/firedoor.dm]: Should check whether it's still in hands instead.

File: code/game/machinery/doors/firedoor.dm [hunk 6]
```diff
@@ -501,7 +512,70 @@ var/global/list/alert_overlays_global = list()
...
+				return 1
+			var/list/directions = new/list(cardinal)
```
Comment (PJB3005) [inline code/game/machinery/doors/firedoor.dm]: What the unholy shit is this?

Just make it go by dir and check only once if that dir is blocked. This is so overengineered.

File: code/game/machinery/doors/firedoor.dm [hunk 7]
```diff
@@ -501,7 +512,70 @@ var/global/list/alert_overlays_global = list()
...
+			for(var/direction in list( user.dir, turn(user.dir,90), turn(user.dir,180), turn(user.dir,270) ))
+				var/found = 0
+				for(var/obj/machinery/door/firedoor/border_only/B in user.loc)
+					if(B.dir == direction)
+						found = 1
+				if(!found)
+					dir_to_set = direction
+					break
+			var/obj/machinery/door/firedoor/border_only/B = new(user.loc,0)
+			B.dir = dir_to_set
```
Comment (PJB3005) [inline code/game/machinery/doors/firedoor.dm]: `B.change_dir(dir_to_set)`

File: code/game/machinery/doors/firedoor.dm [hunk 8]
```diff
@@ -501,7 +512,70 @@ var/global/list/alert_overlays_global = list()
...
+		return 0
+	switch(alert("firedoor construction", "Would you like to construct a full tile firedoor or one direction?", "One Direction", "Full Firedoor", "Cancel", null))
+		if("One Direction")
+			if(!src)
+				return 1
+			if(src.loc != user)
+				return 1
+			var/list/directions = new/list(cardinal)
+			var/i = 0
+			for (var/obj/machinery/door/firedoor/B in user.loc)
```
Comment (PJB3005) [inline code/game/machinery/doors/firedoor.dm]: Make this use `get_turf`. I'm pretty sure right now you could literally build firelocks inside lockers.

File: code/game/machinery/doors/firedoor.dm [hunk 9]
```diff
@@ -501,7 +512,70 @@ var/global/list/alert_overlays_global = list()
...
+			var/dir_to_set = 2
+			for(var/direction in list( user.dir, turn(user.dir,90), turn(user.dir,180), turn(user.dir,270) ))
+				var/found = 0
+				for(var/obj/machinery/door/firedoor/border_only/B in user.loc)
+					if(B.dir == direction)
+						found = 1
+				if(!found)
+					dir_to_set = direction
+					break
+			var/obj/machinery/door/firedoor/border_only/B = new(user.loc,0)
```
Comment (PJB3005) [inline code/game/machinery/doors/firedoor.dm]: `get_turf`

File: code/game/machinery/doors/firedoor.dm [hunk 10]
```diff
@@ -501,7 +512,70 @@ var/global/list/alert_overlays_global = list()
...
+		if("Full Firedoor")
+			if(!src)
+				return 1
+			if(src.loc != user)
+				return 1
+			if(locate(/obj/machinery/door/firedoor) in user.loc)
+				to_chat(user, "<span class='warning'>There is a firedoor already here.</span>")
+				return 1
+			var/obj/machinery/door/firedoor/F = new(user.loc,0)
+			F.dir = SOUTHWEST
```
Comment (PJB3005) [inline code/game/machinery/doors/firedoor.dm]: W-what?

File: code/game/machinery/doors/firedoor.dm [hunk 11]
```diff
@@ -237,6 +239,15 @@ var/global/list/alert_overlays_global = list()
 		force_open(user, C)
 		return
 
+	if(istype(C, /obj/item/weapon/wrench/socket) && blocked)
```
Comment (PJB3005) [inline code/game/machinery/doors/firedoor.dm]: Helpful error message if it isn't welded yet would be appreciated.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Comment by SonixApache** [comment]
a new age of advanced shittery is upon us

File: code/game/machinery/doors/firedoor.dm (hunk 6 above)
Comment (ghost) [reply code/game/machinery/doors/firedoor.dm]: This is literally window placement code.

Two birds with one stone, we make a good way of doing firedoors, and we make better window code

**Comment by ghost** [comment]
Right, border firedoor construction just checks if there is a firedoor facing that direction.

Turns out there is already a check to see if two border firedoors are on the same tile which would qdel the door, made that now drop the frame as well.

**Comment by SonixApache** [comment]
Out of the scope of this PR, but should consider adding post-emag repair functionality.

**Comment by ghost** [comment]
Looking at it, emagging sets blocked to TRUE.

So, repair would be just deconstructing and reconstructing the firedoor with a socket wrench.

**Comment by SonixApache** [comment]
I found a firelock frame outside the box bridge on my local test serb, ~~~will test more.~~~
image
Seems to show up every time. May want to investigate.

**Pull request merged**